questionpy_server/cache.py

```diff
@@ -111,34 +127,35 @@ def put(self, key: str, value: bytes) -> Path:
             raise SizeError(f"Item itself exceeds maximum allowed size of {self.max_bytes} bytes",
                             max_size=self.max_bytes, actual_size=size)
 
-        # Update `_total_bytes` depending on whether the key existed already or not.
-        if key in self._files:
-            self._total_bytes -= self._files[key].size
-        self._total_bytes += size
+        async with self._lock:
+            # Save the bytes on filesystem.
+            path = self.directory / (key + self._extension)
+            tmp_path = path.parent / (path.name + self._tmp_extension)
 
-        # Save the bytes on filesystem.
-        path = self.directory / (key + self._extension)
+            if size != await to_thread(tmp_path.write_bytes, value):
+                raise IOError("Failed to write bytes to file")
 
-        if not self.directory.is_dir():
-            self.directory.mkdir(parents=True, exist_ok=True)
+            await to_thread(tmp_path.rename, path)
 
-        if size != path.write_bytes(value):
-            raise IOError("Failed to write bytes to file")
+            # Update `_total_bytes` depending on whether the key existed already or not.
+            if key in self._files:
+                self._total_bytes -= self._files[key].size
+            self._total_bytes += size
 
-        # Update internal file dictionary.
-        self._files[key] = File(path, size)
+            # Update internal file dictionary.
+            self._files[key] = File(path, size)
 
-        # If size is too large now, remove items until it is less than or equal to the defined maximum.
-        while self._total_bytes > self.max_bytes:
-            # Delete the current oldest item, by instantiating an iterator over all keys (in order)
-            # and passing its next item (i.e. the first one in order) to self.remove.
-            try:
-                self.remove(next(iter(self._files)))
-            except FileNotFoundError:
-                # The file was deleted from the filesystem.
-                pass
+            # If size is too large now, remove items until it is less than or equal to the defined maximum.
+            while self._total_bytes > self.max_bytes:
+                # Delete the current oldest item, by instantiating an iterator over all keys (in order)
+                # and passing its next item (i.e. the first one in order) to self.remove.
+                try:
+                    await self.remove(next(iter(self._files)))
+                except FileNotFoundError:
```

Comment:
Versuche auch bitte mal hier in den Tests die `remove` Methode zu mocken, damit auch getestet wird, ob `FileNotFoundError` abgefangen wird. Dann erreichst du für den cache 100% test coverage. :)